src/scheduler/MpiWorld.cpp

```diff
@@ -379,37 +379,32 @@ std::string MpiWorld::getHostForRank(int rank)
     return host;
 }
 
-std::vector<int> MpiWorld::getRanksForHost(const std::string& host)
+// Group the ranks to host vector by host
+void MpiWorld::setRanksForHost()
 {
-    assert(rankHosts.size() == size);
+    assert(hostForRank.size() == size);
 
-    std::vector<int> ranksForHost;
-    for (int i = 0; i < rankHosts.size(); i++) {
-        if (rankHosts.at(i) == host) {
-            ranksForHost.push_back(i);
-        }
+    for (int rank = 0; rank < hostForRank.size(); rank++) {
+        std::string host = hostForRank.at(rank);
+        ranksForHost[host].push_back(rank);
     }
-
-    return ranksForHost;
 }
 
 // The local leader for an MPI world is defined as the lowest rank assigned to
-// this host
+// this host. For simplicity, we set the local leader to be the first element
+// in the ranks to hosts map.
 void MpiWorld::initLocalRemoteLeaders()
 {
-    std::set<std::string> uniqueHosts(rankHosts.begin(), rankHosts.end());
+    setRanksForHost();
 
-    for (const std::string& host : uniqueHosts) {
-        auto ranksInHost = getRanksForHost(host);
-        // Persist the ranks that are colocated in this host for further use
-        if (host == thisHost) {
-            localRanks = ranksInHost;
-            localLeader =
-              *std::min_element(ranksInHost.begin(), ranksInHost.end());
-        } else {
-            remoteLeaders.push_back(
-              *std::min_element(ranksInHost.begin(), ranksInHost.end()));
+    for (auto it : ranksForHost) {
+        // Persist the local leader in this host for further use
+        if (it.first == thisHost) {
+            localLeader = *std::min_element(it.second.begin(), it.second.end());
```

Comment:
As mentioned in the PR description, I keep the `localLeader` variable to avoid having to write `ranksForHost[thisHost].front()`, but I could do so.